### PR TITLE
Add type annotations

### LINKED ISSUE #653: initiate typing checks
- [ ] tox.ini setup
- [ ] CI with type checking at least some files if cannot be done through out at once. thus marking easy since not full type annotation is requested

<details>
<summary>current state of running mypy</summary> 

```shell
❯ git grep mypy
❯ mypy heudiconv
heudiconv/heuristics/multires_7Tbold.py:25: error: Cannot find implementation or library stub for module named "dicom"  [import]
heudiconv/external/pydicom.py:8: error: Cannot find implementation or library stub for module named "pydicom"  [import]
heudiconv/external/pydicom.py:11: error: Cannot find implementation or library stub for module named "dicom"  [import]
heudiconv/due.py:55: error: Skipping analyzing "duecredit": module is installed, but missing library stubs or py.typed marker  [import]
heudiconv/cli/monitor.py:12: error: Cannot find implementation or library stub for module named "inotify.adapters"  [import]
heudiconv/cli/monitor.py:12: note: See https://mypy.readthedocs.io/en/stable/running_mypy.html#missing-imports
heudiconv/cli/monitor.py:12: error: Cannot find implementation or library stub for module named "inotify"  [import]
heudiconv/cli/monitor.py:13: error: Cannot find implementation or library stub for module named "inotify.constants"  [import]
heudiconv/cli/monitor.py:15: error: Cannot find implementation or library stub for module named "tinydb"  [import]
heudiconv/utils.py:19: error: Cannot find implementation or library stub for module named "nipype.utils.filemanip"  [import]
heudiconv/utils.py:53: error: List or tuple literal expected as the second argument to "namedtuple()"  [misc]
heudiconv/tests/utils.py:61: error: Skipping analyzing "datalad": module is installed, but missing library stubs or py.typed marker  [import]
heudiconv/heuristics/reproin.py:159: error: Need type annotation for "fix_accession2run" (hint: "fix_accession2run: Dict[<type>, <type>] = ...")  [var-annotated]
heudiconv/heuristics/reproin.py:170: error: Need type annotation for "protocols2fix" (hint: "protocols2fix: Dict[<type>, <type>] = ...")  [var-annotated]
heudiconv/heuristics/reproin.py:197: error: Need type annotation for "dicoms2skip" (hint: "dicoms2skip: List[<type>] = ...")  [var-annotated]
heudiconv/external/dcmstack.py:27: error: Cannot find implementation or library stub for module named "dcmstack"  [import]
heudiconv/dicoms.py:24: error: Cannot find implementation or library stub for module named "nibabel.nicom.dicomwrappers"  [import]
heudiconv/dicoms.py:24: error: Cannot find implementation or library stub for module named "nibabel.nicom"  [import]
heudiconv/dicoms.py:24: error: Cannot find implementation or library stub for module named "nibabel"  [import]
heudiconv/dicoms.py:365: error: Missing return statement  [return]
heudiconv/dicoms.py:548: error: Cannot find implementation or library stub for module named "nipype"  [import]
heudiconv/dicoms.py:606: error: Cannot find implementation or library stub for module named "dcmstack.extract"  [import]
heudiconv/dicoms.py:606: error: Cannot find implementation or library stub for module named "dcmstack"  [import]
heudiconv/external/dlad.py:18: error: Skipping analyzing "datalad.api": module is installed, but missing library stubs or py.typed marker  [import]
heudiconv/external/dlad.py:37: error: Skipping analyzing "datalad": module is installed, but missing library stubs or py.typed marker  [import]
heudiconv/external/dlad.py:39: error: Skipping analyzing "datalad.support.annexrepo": module is installed, but missing library stubs or py.typed marker  [import]
heudiconv/external/dlad.py:40: error: Skipping analyzing "datalad.support.external_versions": module is installed, but missing library stubs or py.typed marker  [import]
heudiconv/bids.py:618: error: Cannot find implementation or library stub for module named "nibabel"  [import]
heudiconv/convert.py:670: error: Cannot find implementation or library stub for module named "nipype"  [import]
heudiconv/convert.py:675: error: Cannot find implementation or library stub for module named "nipype.interfaces.dcm2nii"  [import]
heudiconv/convert.py:733: error: Cannot find implementation or library stub for module named "nipype.interfaces.base"  [import]
heudiconv/main.py:299: error: Skipping analyzing "etelemetry": module is installed, but missing library stubs or py.typed marker  [import]
heudiconv/tests/test_regression.py:17: error: Skipping analyzing "datalad.support.exceptions": module is installed, but missing library stubs or py.typed marker  [import]
heudiconv/tests/test_monitor.py:7: error: Cannot find implementation or library stub for module named "tinydb"  [import]
heudiconv/tests/test_monitor.py:16: error: First argument to namedtuple() should be "Header", not "header"  [name-match]
heudiconv/tests/test_monitor.py:71: error: Argument 1 to "CalledProcessError" has incompatible type "str"; expected "int"  [arg-type]
heudiconv/tests/test_monitor.py:71: error: Argument 2 to "CalledProcessError" has incompatible type "int"; expected "Union[Union[str, bytes, PathLike[str], PathLike[bytes]], Sequence[Union[str, bytes, PathLike[str], PathLike[bytes]]]]"  [arg-type]
heudiconv/tests/test_main.py:121: error: Skipping analyzing "datalad.api": module is installed, but missing library stubs or py.typed marker  [import]
heudiconv/tests/test_heuristics.py:25: error: Skipping analyzing "datalad.api": module is installed, but missing library stubs or py.typed marker  [import]
heudiconv/tests/test_bids.py:20: error: Cannot find implementation or library stub for module named "nibabel"  [import]
heudiconv/tests/test_bids.py:59: error: Skipping analyzing "datalad.support.exceptions": module is installed, but missing library stubs or py.typed marker  [import]
Found 40 errors in 18 files (checked 48 source files)

```
</details>

### PR DESCRIPTION
**Pull request by jwodder**
Closes #653.

---

Problems encountered so far with applying type annotations:

* `heudiconv.dicoms.group_dicoms_into_seqinfos()`: If `grouping == "custom"` and `custom_grouping` is a callable, the return value is the result of applying `custom_grouping()` to some arguments; otherwise, the return value is something else.  Normally, this could be annotated by using overloads and `Literal["custom"]`, but whenever `group_dicoms_into_seqinfos()` is called in the heudiconv code, `grouping` (if present) is always passed as a variable rather than a string literal, and so `Literal` is unusable here.
* The docstring for `heudiconv.dicoms.create_seqinfo()` states that the first argument is of type `nibabel.nicom.dicomwrappers.MosaicWrapper`, yet [here](https://github.com/nipy/heudiconv/blob/012895dfc1d940ffa42213e0fec15d8392f5b7f3/heudiconv/dicoms.py#L280) the supplied argument (obtained from `validate_dicom()`) is just a `nibabel.nicom.dicomwrappers.Wrapper`.
* `SeqInfo.series_id` is clearly meant to be a `str`, yet there are several places in `heudiconv/heuristics/example.py` where this field is compared against or assigned to an `int` variable.

### CODE REVIEW

**Comment by yarikoptic** [comment]
I just merged a small PR which introduced conflicts, please resolve.

**Comment by jwodder** [comment]
@nipy/team-heudiconv Yet another issue: In the following code:

https://github.com/nipy/heudiconv/blob/2c6d228516dda41cdc9f91e7bea51b3acdb0f5b7/heudiconv/bids.py#L711-L717

the `header.get_best_affine()` and `header.get_data_shape()` methods are not present on all possible return types of `nibabel.load()`; as far as I can tell, they are only present when `nibabel.load()` returns an instance of one of the following types:

* `Nifti1Pair`
* `Nifti1Image`
* `AnalyzeImage`
* `MGHImage`

Exactly what type is `nb_load(nifti_file)` expected to return here?

**Comment by jwodder** [comment]
@nipy/team-heudiconv This should hopefully be the last typing issue I post: The `prov_file` argument to `heudiconv.dicoms.embed_metadata_from_dicoms()` can apparently only be `str`, yet in [this call to the function](https://github.com/nipy/heudiconv/blob/5e6c7ea6f2b5fb91a123fa52890a8c1d5e2cc725/heudiconv/convert.py#L627), the `prov_file` argument can be either a `str` or `None`.  What should be done about this?

**Comment by jwodder** [comment]
@nipy/team-heudiconv Ping.

**Comment by yarikoptic** [comment]
First of all thank you @jwodder for going on this types quest!

> * I believe I may have found a bug, and I'm not sure how to address it.  I've determined that the `ses` argument to `heudiconv.convert.prep_conversion()` is of type `str | int | None` (See explanation below), yet [here](https://github.com/nipy/heudiconv/blob/2c6d228516dda41cdc9f91e7bea51b3acdb0f5b7/heudiconv/convert.py#L124-L125) `ses` is passed to `sanitize_label()`, which only accepts `str`s and will error if given an `int`.  What should be done about this?

`ses` is in general a "label", not an "index" (see [entity table](https://bids-specification.readthedocs.io/en/stable/appendices/entity-table.html) for more info on those) so should be a `str`. So I guess the point at which you figured it could come out as an `int` should format it to `str` one way or another.

<details>
<summary>So as you deduced -- we should just just tune up that single place where it is used as `int`, so may be that is the initial diff/the point where used as `int`:</summary> 

```diff
diff --git a/heudiconv/parser.py b/heudiconv/parser.py
index 354b160..7dd73f1 100644
--- a/heudiconv/parser.py
+++ b/heudiconv/parser.py
@@ -105,7 +105,7 @@ def get_extracted_dicoms(fl: Iterable[str]) -> ItemsView[Optional[int], list[str
     the unarchived file. If there are multiple archived files, they are grouped
     into separate sessions.
     """
-    sessions: dict[Optional[int], list[str]] = defaultdict(list)
+    sessions: dict[Optional[str], list[str]] = defaultdict(list)
 
     # keep track of session manually to ensure that the variable is bound
     # when it is used after the loop (e.g., consider situation with
@@ -136,14 +136,14 @@ def get_extracted_dicoms(fl: Iterable[str]) -> ItemsView[Optional[int], list[str
             os.chmod(f, mode=0o700)
         # store full paths to each file, so we don't need to drag along
         # tmpdir as some basedir
-        sessions[session] = archive_content
+        sessions[str(session)] = archive_content
         session += 1
 
     if session == 1:
         # we had only 1 session (and at least 1), so not really multiple
         # sessions according to classical 'heudiconv' assumptions, thus
         # just move them all into None
-        sessions[None] += sessions.pop(0)
+        sessions[None] += sessions.pop("0")
 
     return sessions.items()
 
diff --git a/heudiconv/utils.py b/heudiconv/utils.py
index 7284479..37a1237 100644
--- a/heudiconv/utils.py
+++ b/heudiconv/utils.py
@@ -78,7 +78,7 @@ class StudySessionInfo(NamedTuple):
     # StudyInstanceUID into some place within hierarchy
     locator: Optional[str]
 
-    session: Optional[str | int]
+    session: Optional[str]
 
     # should be some ID defined either in cmdline or deduced
     subject: Optional[str]

```
</details>


> Some of the code in `heudiconv/convert.py` extracts fields from BIDS JSON (sidecar?) files,

correct ! called "sidecar" metadata files in BIDS

>  yet I cannot determine what types these fields are supposed to be. Specifically, the fields in question are:
> 
> * `CoilString` — Not found on https://bids-specification.readthedocs.io

some times we might be "ahead of the standard" here and get a better luck searching directly in https://github.com/bids-standard/bids-specification . This time it leads to unfortunately never finished (still open) https://github.com/bids-standard/bids-specification/pull/425 . From https://github.com/bids-standard/bids-specification/pull/425/files#diff-37df75f7d2d341bed5815aad788fe7bff09374c335eb0c25fab0b7569e8c0a46R2 - just a `str`

> * `EchoNumber` — Not found on https://bids-specification.readthedocs.io

hm, nothing among PRs. Not standardized within BIDS, but extracted by dcm2niix

```shell
❯ git grep EchoNumber
BIDS/README.md:| EchoNumber                         |      | Only multi-echo series                                                                  | D          |
GE/README.md:Current GE software (DV26.0_R03_1831.b) running research multi-echo sequences create invalid DICOM images. The required public [EchoTime (0018,0081)](https://dicom.innolitics.com/ciods/mr-image/mr-image/00180081) attribute lists the shortest echo time for the series, rather than the actual echo time for the given DICOM image. The public tag [EchoNumber (0018,0086)](https://dicom.innolitics.com/ciods/mr-image/mr-image/00180086) reports `1` for all echoes. These limitations in GE's DICOM images disrupt dcm2niix's image conversion. Hopefully future product sequences will generate valid DICOM data. In the meantime, [issue 359](https://github.com/rordenlab/dcm2niix/issues/359) provides a kludge for image conversion.
console/nii_dicom_batch.cpp:                fprintf(fp, "\t\"EchoNumber\": %d,\n", d.echoNum);
```

should be an `int`. @pvelasco @tsalo - should we submit a PR for BIDS to formalize it?

> * `EchoTime` — The BIDS spec's "Glossary of schema objects" [lists four fields by that name](https://bids-specification.readthedocs.io/en/stable/glossary.html#echotime-sense-1-metadata)

to say the truth I am not yet sure about plurality of definitions there, but it should be a `float`. I don't think we should worry about possible `list[float]` here

> * `ImageType` — Not found on https://bids-specification.readthedocs.io

yet another one `dcm2niix` extracts from DICOMs, but not standardized in BIDS (not sure if this one should, but also might): https://github.com/rordenlab/dcm2niix//blob/HEAD/BIDS/README.md#global-series-information 

Make it `list[str]`.

> @nipy/team-heudiconv Also, side question: In this code:
> 
> https://github.com/nipy/heudiconv/blob/2c6d228516dda41cdc9f91e7bea51b3acdb0f5b7/heudiconv/tests/test_dicoms.py#L30-L33
> 
> is `dcm.pub` supposed to be `getattr(dcm, pub)`? That would seem to make more sense.

well spotted!!!

**Comment by yarikoptic** [comment]
> * `Nifti1Pair`
> * `Nifti1Image`
> * `AnalyzeImage`
> * `MGHImage`
> 
> Exactly what type is `nb_load(nifti_file)` expected to return here?

given our invocation of dcm2niix I believe it is exclusively `Nifti1Image`

> The `prov_file` argument to `heudiconv.dicoms.embed_metadata_from_dicoms()` can apparently only be `str`

why? I think it could still be `None` if e.g. `with_prov` was False and not provided to `convert`. So probably should just be `str | None` ?

**Comment by jwodder** [comment]
@yarikoptic I still need a resolution for the issues mention in [my top comment](https://github.com/nipy/heudiconv/pull/656#issue-1646168126).

**Comment by jwodder** [comment]
@yarikoptic Specifying that the `custom_grouping` argument to `group_dicoms_into_seqinfos()` must return `dict[SeqInfo, list[str]]` doesn't solve the problem, as `group_dicoms_into_seqinfos()` can return either that or `dict[Optional[str], dict[SeqInfo, list[str]]]` depending on the value of `flatten` — yet the function is called with a `custom_grouping` and `flatten=True` [here](https://github.com/nipy/heudiconv/blob/502bf49d9532d82546283fe31a5f1398ba3acc66/heudiconv/convert.py#L189), and it's called with a `custom_grouping` and `flatten=False` [here](https://github.com/nipy/heudiconv/blob/502bf49d9532d82546283fe31a5f1398ba3acc66/heudiconv/parser.py#L205).  The only way for mypy to be sure what type it's getting would be if the values of `flatten`, `grouping`, and `custom_grouping` were all statically known.

**Comment by yarikoptic** [comment]
but it seems that we can't know them statically and moreover we "violate" it in case of `flatten=False` whenever `Callable` `custom_groupping` is provided.  So -- may be for now just comment out the `@overload`s you defined for the `group_dicoms_into_seqinfos` and add a comment that such clear/strict typing is not needed since `flatten` is ignored in case of `Callable` `custom_groupping` which is expected to return the other kind?

**Comment by yarikoptic** [comment]
FWIW -- just checked that it would not be a solution since then code which calls `group_dicoms_into_seqinfos` would not be able to tell one or another...

**Comment by yarikoptic** [comment]
I've tuned up that `example.py`, fixed another minor bug (can go outside of this PR in principle) and also added that casting

```
-            return custom_grouping(files, dcmfilter, SeqInfo)
+            return cast(Dict[SeqInfo, List[str]],
+                        custom_grouping(files, dcmfilter, SeqInfo))
```
and `mypy` seems to be happy for me locally, so may be cast'ing is just enough here ?

**Comment by yarikoptic** [comment]
hm, on CI (where we have py 3.7) it still fails with following

```
heudiconv/convert.py:55: error: Cannot find implementation or library stub for
module named "type_extensions"  [import]
            from type_extensions import TypedDict
    ^
heudiconv/convert.py:55: note: See https://mypy.readthedocs.io/en/stable/running_mypy.html#missing-imports
heudiconv/convert.py:57: error: Unexpected keyword argument "total" for
"__init_subclass__" of "object"  [call-arg]
        class PopulateIntendedForOpts(TypedDict, total=False):
        ^
.tox/typing/lib/python3.7/site-packages/mypy/typeshed/stdlib/builtins.pyi:113: note: "__init_subclass__" of "object" defined here
Found 2 errors in 1 file (checked 46 source files)
typing: exit 1 (14.48 seconds) /home/runner/work/heudiconv/heudiconv> mypy heudiconv pid=1840
.pkg: _exit> python /opt/hostedtoolcache/Python/3.7.16/x64/lib/python3.7/site-packages/pyproject_api/_backend.py True setuptools.build_meta
  typing: FAIL code 1 (58.[26](https://github.com/nipy/heudiconv/actions/runs/4812417000/jobs/8567639256?pr=656#step:5:27)=setup[43.78]+cmd[14.48] seconds)
  evaluation failed :( (58.40 seconds)
```
may be we just should go to newer python for type checking?

**Comment by yarikoptic** [comment]
@jwodder please check two last small commits where I have (I think) constrained typing a little more - let me know if may be you have reservations against that.

File: heudiconv/bids.py [hunk 1]
```diff
@@ -103,8 +103,10 @@ def maybe_na(val: Any) -> str:
         return "n/a"
 
 
-def treat_age(age: str|float) -> str:
+def treat_age(age: str | float | None) -> str | None:
     """Age might encounter 'Y' suffix or be a float"""
+    if age is None:
+        return None  # might be converted to N/A by maybe_na
     agestr = str(age)
     if agestr.endswith("M"):
```
Comment (jwodder) [inline heudiconv/bids.py:111]: If `age` is limited to `str | float`, then `if agestr.endswith("M")` will never be true.  Why is this branch here?

Comment (jwodder) [reply heudiconv/bids.py:111]: Disregard this; I somehow read `str` as `int`.

**Comment by yarikoptic** [comment]
ok -- Let's go. Thank you again @jwodder for all this monumental work!

**Comment by yarikoptic** [comment]
although indeed `internal` I want next release to be `minor` version boost at least due to the scale of this change, so labeling it is as such

**Pull request merged**